scripts/write-api-docs.js

```diff
@@ -0,0 +1,943 @@
+#!/usr/bin/env node
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// This script generates API documentation. By default it writes output
+// to `docs/api.md`, but you can override that by specifying a different
+// path as an argument. Requires node 6 or greater.
+
+'use strict'
+
+const acorn = require('acorn')
+const P = require('../lib/promise')
+const fs = P.promisifyAll(require('fs'), { suffix: 'P' })
+const path = require('path')
+const handlebars = require('handlebars')
+const render = handlebars.compile(
+  fs.readFileSync(path.resolve(__dirname, 'api-docs.handlebars'), { encoding: 'utf8' })
+)
+
+// In the API docs, any text inside <!--begin-xxx--> / <!--end-xxx-> delimiters
+// will be propagated to the same place each time the docs are regenerated.
+const DOCS_DATA = /<!--begin-([a-z_]+(?:-[a-z_]+)*)-->((?:.*?\n?.*?)*?)<!--end-([a-z_]+(?:-[a-z_]+)*)-->/gi
+const ACORN_OPTIONS = {
+  locations: true,
+  sourceType: 'script'
+}
+const EMPTY_MAP = new Map()
+const IGNORE = new Set([ 'defaults.js', 'idp.js', 'index.js', 'validators.js' ])
+// HACK: Assumes the location of lib/routes/*.js
+const ROUTES_DIR = path.resolve(__dirname, '../lib/routes')
+const FUNCTION_EXPRESSION_TYPES = new Set([ 'FunctionExpression', 'ArrowFunctionExpression' ])
+const ARRAY_TYPES = new Set([ 'ArrayExpression' ])
+const RETURN_TYPES = new Set([ 'ReturnStatement' ])
+const OBJECT_TYPES = new Set([ 'ObjectExpression' ])
+const LITERAL_TYPES = new Set([ 'Literal' ])
+const CALL_TYPES = new Set([ 'CallExpression' ])
+const MEMBER_TYPES = new Set([ 'MemberExpression' ])
+const ERROR_PROPERTY_TYPES = new Set([ 'Literal', 'MemberExpression', 'BinaryExpression', 'LogicalExpression' ])
+const SESSION_TOKEN_STRATEGY = /^sessionToken/
+const KEY_FETCH_TOKEN_STRATEGY = /^keyFetchToken/
+const NOT_ERRORS = new Set([ 'toString', 'header', 'backtrace', 'translate' ])
+
+const args = parseArgs()
+
+Promise.all([
+  parseDocs(args.path),
+  parseErrors()
+])
+.then(([ docs, errors ]) => Promise.all([
```

Comment:
>Do you want me to change it?

Yeah, we are still on node 4 in all repos, so we need to be able to generate docs from that :)